podium-gateway/src/main/java/nl/thehyve/podium/service/DeliveryProcessService.java

```diff
@@ -0,0 +1,216 @@
+        // Update the delivery process object with the process variables.
+        HistoricProcessInstance instance = findProcessInstance(deliveryProcess.getProcessInstanceId());
+        Map<String, Object> variables = instance.getProcessVariables();
+        deliveryProcess.setStatus((DeliveryStatus) variables.get("status"));
+        deliveryProcess.setOutcome((DeliveryProcessOutcome) variables.get("outcome"));
+        deliveryProcess = deliveryProcessRepository.save(deliveryProcess);
+        // save to elastic search as well
+        deliveryProcessSearchRepository.save(deliveryProcess);
+        return deliveryProcess;
+    }
+
+    private DeliveryProcess completeCurrentTask(AuthenticatedUser user, DeliveryProcess deliveryProcess) {
+        DeliveryStatus status = deliveryProcess.getStatus();
+        if (status == DeliveryStatus.None) {
+            throw new RuntimeException("Invalid status.");
+        }
+        DeliveryTask deliveryTask = taskForStatus.get(status);
+        Task task = findTaskByRequestId(deliveryProcess.getProcessInstanceId(), deliveryTask.getTaskId());
+        if (task == null) {
+            throw new ResourceNotFound("Task not found for status " + status.name() + " (" + deliveryTask.getTaskId() + ")");
+        }
+        taskService.addUserIdentityLink(task.getId(), user.getUuid().toString(), IdentityLinkType.PARTICIPANT);
+        if (task.getDelegationState() == DelegationState.PENDING) {
+            taskService.resolveTask(task.getId());
+        }
+        taskService.complete(task.getId());
+
+        return updateDeliveryProcess(deliveryProcess);
+    }
+
+    private void setVariable(DeliveryProcess deliveryProcess, DeliveryVariable variable, Object value) {
+        runtimeService.setVariable(
+            deliveryProcess.getProcessInstanceId(),
+            variable.getVariableName(),
+            value);
+    }
+
+    public DeliveryProcess release(AuthenticatedUser user, DeliveryProcess deliveryProcess) throws ActionNotAllowedInStatus {
+        if (deliveryProcess.getStatus() == DeliveryStatus.Preparation) {
+            setVariable(deliveryProcess, DeliveryVariable.Release, Boolean.TRUE);
+            return completeCurrentTask(user, deliveryProcess);
+        }
+        throw ActionNotAllowedInStatus.forStatus(deliveryProcess.getStatus());
+    }
+
+    public DeliveryProcess received(AuthenticatedUser user, DeliveryProcess deliveryProcess) throws ActionNotAllowedInStatus {
+        if (deliveryProcess.getStatus() == DeliveryStatus.Released) {
+            setVariable(deliveryProcess, DeliveryVariable.Received, Boolean.TRUE);
+            return completeCurrentTask(user, deliveryProcess);
+        }
+        throw ActionNotAllowedInStatus.forStatus(deliveryProcess.getStatus());
+    }
+
+    public DeliveryProcess cancel(AuthenticatedUser user, DeliveryProcess deliveryProcess) throws ActionNotAllowedInStatus {
+        if (deliveryProcess.getStatus() == DeliveryStatus.Released) {
```

Comment:
You should also be able to cancel a deliveryProcess when its not in 'Released'. This is to ensure maximum flexibility of the process flow and support cancellation of deliveries also when they're being 'Prepared'.